share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -0,0 +1,533 @@
+                            "write-share-group-state",
+                            topicPartitionFor(SharePartitionKey.getInstance(groupId, topicData.topicId(), partitionData.partition())),
+                            Duration.ofMillis(config.shareCoordinatorWriteTimeoutMs()),
+                            coordinator -> coordinator.writeState(context, new WriteShareGroupStateRequestData()
+                                .setGroupId(groupId)
+                                .setTopics(Collections.singletonList(new WriteShareGroupStateRequestData.WriteStateData()
+                                    .setTopicId(topicData.topicId())
+                                    .setPartitions(Collections.singletonList(new WriteShareGroupStateRequestData.PartitionData()
+                                        .setPartition(partitionData.partition())
+                                        .setStartOffset(partitionData.startOffset())
+                                        .setLeaderEpoch(partitionData.leaderEpoch())
+                                        .setStateEpoch(partitionData.stateEpoch())
+                                        .setStateBatches(partitionData.stateBatches())))))))
+                        .exceptionally(exception -> handleOperationException(
+                            "write-share-group-state",
+                            request,
+                            exception,
+                            (error, message) -> WriteShareGroupStateResponse.toErrorResponseData(
+                                topicData.topicId(),
+                                partitionData.partition(),
+                                error,
+                                "Unable to write share group state: " + exception.getMessage()
+                            ),
+                            log
+                        ));
+                    partitionFut.put(partitionData.partition(), future);
+                });
+        });
+
+        // Combine all futures into a single CompletableFuture<Void>
+        CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futureMap.values().stream()
+            .flatMap(partMap -> partMap.values().stream()).toArray(CompletableFuture[]::new));
+
+        // topicId -> {partitionId -> responseFuture}
+        return combinedFuture.thenApply(v -> {
+            List<WriteShareGroupStateResponseData.WriteStateResult> writeStateResults = new LinkedList<>();
+            futureMap.forEach(
+                (topicId, topicEntry) -> {
+                    List<WriteShareGroupStateResponseData.PartitionResult> partitionResults = new LinkedList<>();
+                    topicEntry.forEach(
+                        // map of partition id -> responses from api
+                        (partitionId, responseFut) -> {
+                            long timeTaken = time.hiResClockMs() - startTime;
```

Comment:
this is an extremely pedantic nitpick, but we should get the current time once outside of the loop. otherwise, the time delta calculation will be somewhat skewed by the execution time of the loop